core/include/scipp/core/values_and_variances_strong_types.h

```diff
@@ -0,0 +1,202 @@
+
+template <class... Args> struct Values : ValuesTaggedTupple<Args...> {
+  using tag_type = detail::ValuesTag;
+  Values(Args &&... args)
+      : ValuesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(std::forward<Args>(args)...)) {}
+  template <class T>
+  Values(std::initializer_list<T> init)
+      : ValuesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(init.begin(), init.end())) {}
+};
+template <class... Args> Values(Args &&... args)->Values<Args...>;
+template <class T>
+Values(std::initializer_list<T>)
+    ->Values<typename std::initializer_list<T>::iterator,
+             typename std::initializer_list<T>::iterator>;
+
+template <class... Args>
+using VariancesTaggedTupple =
+    detail::MakeTaggedTupleReturnType<detail::VariancesTag, Args...>;
+
+template <class... Args> struct Variances : VariancesTaggedTupple<Args...> {
+  using tag_type = detail::VariancesTag;
+  Variances(Args &&... args)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(std::forward<Args>(args)...)) {}
+  template <class T>
+  Variances(std::initializer_list<T> init)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(init.begin(), init.end())) {}
+};
+template <class... Args> Variances(Args &&... args)->Variances<Args...>;
+template <class T>
+Variances(std::initializer_list<T>)
+    ->Variances<typename std::initializer_list<T>::iterator,
+                typename std::initializer_list<T>::iterator>;
+
+namespace detail {
+template <class Tag, class T> struct has_tag : std::false_type {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Values<Ts...>> : std::is_same<Tag, ValuesTag> {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Variances<Ts...>> : std::is_same<Tag, VariancesTag> {};
+
+template <class T, class... Args>
+constexpr bool is_type_in_pack_v =
+    std::disjunction<std::is_same<T, std::decay_t<Args>>...>::value;
+
+template <class Tag, class... Args>
+constexpr bool is_tag_in_pack_v =
+    std::disjunction<has_tag<Tag, Args>...>::value;
+
+template <class T, template <class T1, class T2> class Cond, class... Args>
+class Indexer {
+  template <std::size_t... IS>
+  static constexpr auto indexOfCorresponding_impl(std::index_sequence<IS...>) {
+    return ((Cond<T, Args>::value * IS) + ...);
+  }
+
+public:
+  static constexpr auto indexOfCorresponding() {
+    return indexOfCorresponding_impl(
+        std::make_index_sequence<sizeof...(Args)>{});
+  }
+};
+
+template <class T, class Tuple, std::size_t... I>
+constexpr T make_move_from_tuple_impl(Tuple &&t, std::index_sequence<I...>) {
+  return T(std::move(std::get<I>(t))...);
+}
+
+template <class T, class Tuple> constexpr T make_move_from_tuple(Tuple &&t) {
+  return detail::make_move_from_tuple_impl<T>(
+      std::forward<Tuple>(t),
+      std::make_index_sequence<
+          std::tuple_size_v<std::remove_reference_t<Tuple>>>{});
```

Comment:
It looks like copy and paste from `std::make_from_tuple` apart from the added `std::move` and remove `std::forward`? Is the problem that our `t` is not actually an `rvalue` reference? It looks like you could move at the call site, so I do not understand why not.

See https://en.cppreference.com/w/cpp/utility/make_from_tuple.